src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp

```diff
@@ -14,241 +14,513 @@
+        bool status = resolve_auto_pad_for_shape(op, pads_begin, pads_end, input_shapes, 2, is_grouped ? 3 : 2);
+        OPENVINO_ASSERT(status,
+                        "Convolution shape inference doesn't have enough information to calculate static shapes");
+        shape_infer(op, pads_begin, pads_end, input_shapes, output_shapes);
+        return output_shapes;
+    }
+
+protected:
+    ov::CoordinateDiff pads_begin, pads_end;
+    bool is_grouped;
+};
+
+template <typename OP>
+class entryConvBackprop : public entryBase {
+public:
+    entryConvBackprop(std::shared_ptr<OP> node, bool is_grouped) : entryBase(node), is_grouped(is_grouped) {}
+    const ov::CoordinateDiff& get_pads_begin() override {
+        return pads_begin;
+    }
+    const ov::CoordinateDiff& get_pads_end() override {
+        return pads_end;
+    }
+    std::vector<ov::StaticShape> infer(
+        const std::vector<ov::StaticShape>& input_shapes,
+        const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data) override {
+        ov::StaticShape output_shape_input;
+        auto op = static_cast<OP*>(node.get());
+        std::vector<ov::StaticShape> output_shapes(op->get_output_size());
+        if (op->get_input_size() == 3)
+            get_data_as_shape<ov::StaticShape>(2, op, output_shape_input, constant_data);
+        bool status = resolve_auto_pad_for_shape_back_prop(op,
+                                                           pads_begin,
+                                                           pads_end,
+                                                           input_shapes,
+                                                           output_shape_input,
+                                                           2,
+                                                           is_grouped ? 3 : 2);
+        OPENVINO_ASSERT(
+            status,
+            "ConvolutionBackpropData shape inference doesn't have enough information to calculate static shapes");
+        shape_infer(op, pads_begin, pads_end, output_shape_input, input_shapes, output_shapes);
+        return output_shapes;
+    }
+
+protected:
+    ov::CoordinateDiff pads_begin, pads_end;
+    bool is_grouped;
+};
+
+template <typename OP>
+std::shared_ptr<entryIOC<OP>> make_shared_entryIOC(std::shared_ptr<OP> node) {
+    return std::make_shared<entryIOC<OP>>(node);
+}
+
+template <typename OP>
+std::shared_ptr<entryIO<OP>> make_shared_entryIO(std::shared_ptr<OP> node) {
+    return std::make_shared<entryIO<OP>>(node);
+}
+
+std::shared_ptr<IShapeInfer> make_shape_inference(const std::shared_ptr<ngraph::Node>& op) {
+    if (auto node = ov::as_type_ptr<ov::opset8::Convolution>(op)) {
+        return std::make_shared<entryConv<ov::opset8::Convolution>>(node, false);
+    } else if (auto node = ov::as_type_ptr<ov::opset8::GroupConvolution>(op)) {
+        return std::make_shared<entryConv<ov::opset8::GroupConvolution>>(node, true);
```

Comment:
Do I understand correctly that these dynamic casts `ov::as_type_ptr<ov::opset8::Convolution>` were taking a lot of time in the previous measurements. Were previous measurements wrong? 
I thought we wish to have a hash table of inferences instead.

Reply:
No, these dynamic casts do take considerable time comparing to actual shape inference workload, but now they are executed only once at network loading time rather than at each call of shape inference.

That's because we changed the original API into a OOP form, at MKLDNNNode constructor, make_shape_inference() returns a shared_ptr to an object named `shapeInference` which is already instantiated from a proper subclass which handles shape inference specifically for target op. and it's member function will be called at shape inference time w/o the need for additional type cast.

Reply:
Okay! That sounds great! Thank you for explanation

Reply:
No problem, you are welcome!
This method turns out to be superior comparing to my original proposal of using hash-table, not only in performance, but also allow adding additional interface to pass padding begin/end information to caller.